src/expression.c

```diff
@@ -6595,33 +6595,48 @@ Expression *BinExp::binSemanticProp(Scope *sc)
     return NULL;
 }
 
-Expression *BinExp::checkComplexOpAssign(Scope *sc)
+Expression *BinAssignExp::checkOpAssignTypes(Scope *sc)
 {
+    // At that point t1 and t2 are the merged types. type is the original type of the lhs.
+    Type *t1 = e1->type;
+    Type *t2 = e2->type;
+
+    // T opAssign floating yields a floating. Prevent truncating conversions (float to int).
+    // See issue 3841.
+    // Should we also prevent double to float (type->isfloating() && type->size() < t2 ->size()) ?
+    if (op == TOKmulass || op == TOKdivass || op == TOKmodass || TOKaddass || op == TOKminass || op == TOKpowass)
+    {
+        if ((type->isintegral() && t2->isfloating()))
+        {
+            error("%s %s %s is undefined (refusing to perform truncating conversion)",
```

Comment:
This should be made a warning, because it will force people to change their code. The idea is warning, then deprecation, then finally error, after some major time has elapsed.